src/mbgl/gl/context.cpp

```diff
@@ -124,10 +132,23 @@ void Context::initializeExtensions(const std::function<gl::ProcAddress(const cha
             return nullptr;
         };
 
-        debugging = std::make_unique<extension::Debugging>(fn);
-        if (!disableVAOExtension) {
-            vertexArray = std::make_unique<extension::VertexArray>(fn);
-        }
+        const std::string renderer = reinterpret_cast<const char*>(MBGL_CHECK_ERROR(glGetString(GL_RENDERER)));
+        Log::Info(Event::General, "GPU Identifier: %s", renderer.c_str());
+
+        debugging = initializeExtension<extension::Debugging>(fn, [&renderer]() {
+            // Blacklist ANGLE since the debugging extension is causing crashes
+            return renderer.find("ANGLE") != std::string::npos;
+        });
+        vertexArray = initializeExtension<extension::VertexArray>(fn, [&renderer]() {
+            // Blacklist Adreno 2xx, 3xx as it crashes on glBuffer(Sub)Data
+            // Blacklist ARM Mali-T720 (in some MT8163 chipsets) as it crashes on glBindVertexArray
+            // Blacklist ANGLE as the combination of Qt + Windows + ANGLE leads to crashes
+            return renderer.find("Adreno (TM) 2") != std::string::npos
+                || renderer.find("Adreno (TM) 3") != std::string::npos
+                || renderer.find("ANGLE") != std::string::npos
```

Comment:
Why did you add ANGLE to the VertexArray exclusion as well?

Reply:
I tried to capture that in the first commit's message: https://github.com/mapbox/mapbox-gl-native/pull/12754/commits/d6180d1cbfbfcf30803b2792e54ab3df8dffdf99

The idea is to unify the \#ifdef disabling the extension on Windows with the blocklist. There's an assumption that ANGLE is only being used on Windows, though.

Reply:
I removed the assumption on ANGLE only being used on windows. The extensions are only disabled now if ANGLE is running on Direct3D.